File: Telerik/SecondControl/04.ValidateParanthesis.java
package Telerik;

import java.util.*;

public class ValidateParanthesis {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        int n = Integer.parseInt(scan.nextLine());
        for (int i = 0; i < n; i++) {
            String line = scan.nextLine();
            if (line.equals("")) {
                System.out.println("yes");
                continue;
            }
            Boolean isValid = false;
            ArrayList<String> possibleExpressions = new ArrayList<String>();
            Queue<String> expressions = new LinkedList<String>();
            expressions.add(line);
            while (expressions.size() > 0) {
                String currentExpression = expressions.remove();
                if (currentExpression.contains("*")) {
                    expressions.add(ReplaceFirst(currentExpression, "("));
                    expressions.add(ReplaceFirst(currentExpression, ")"));
                    expressions.add(ReplaceFirst(currentExpression, ""));
                } else {
                    possibleExpressions.add(currentExpression);
                }
            }
            for (int j = 0; j < possibleExpressions.size(); j++) {
                if (isVadidExpression(possibleExpressions.get(j))) {
                    isValid = true;
                    break;
                }
            }
            //System.out.println(String.join("\n", possibleExpressions));
            System.out.println(isValid ? "yes" : "no");
        }
    }

    static String ReplaceFirst(String expression, String replacer) {
        int index = expression.indexOf('*');
        String newExpression = "";
        newExpression = expression.substring(0, index) + replacer + expression.substring(index + 1);
        return newExpression;
    }

    static Boolean isVadidExpression(String expression) {
        Stack<Boolean> validations = new Stack<Boolean>();
        for (int i = 0; i < expression.length(); i++) {
            if (expression.charAt(i) == '(') {
                validations.push(true);
            }
            if (expression.charAt(i) == ')' && validations.size() > 0) {
                validations.pop();
            } else if (expression.charAt(i) == ')' && validations.size() == 0) {
                return false;
            }
        }
        if (validations.size() == 0) {
            return true;
        } else {
            return false;
        }
    }
}

File: Telerik/SecondControl/04.ValidateParanthesis.cs
using System;
using System.Collections.Generic;

class ValidateParanthesis
{
    static void Main()
    {
        int n = int.Parse(Console.ReadLine());
        for (int i = 0; i < n; i++)
        {
            string line = Console.ReadLine();
            if (line == string.Empty)
            {
                Console.WriteLine("yes");
                continue;
            }
            bool isValid = false;
            List<string> possibleExpressions = new List<string>();
            Queue<string> expressions = new Queue<string>();
            expressions.Enqueue(line);
            while (expressions.Count > 0)
            {
                string currentExpression = expressions.Dequeue();
                if (currentExpression.Contains("*"))
                {
                    expressions.Enqueue(ReplaceFirst(currentExpression, "("));
                    expressions.Enqueue(ReplaceFirst(currentExpression, ")"));
                    expressions.Enqueue(ReplaceFirst(currentExpression, ""));
                }
                else
                {
                    possibleExpressions.Add(currentExpression);
                }
            }
            for (int j = 0; j < possibleExpressions.Count; j++)
            {
                if (isVadidExpression(possibleExpressions[j]))
                {
                    isValid = true;
                    break;
                }
            }
            //Console.WriteLine(String.Join("\n", possibleExpressions));
            Console.WriteLine(isValid ? "yes" : "no");
        }
    }

    static string ReplaceFirst(string expression, string replacer)
    {
        int index = expression.IndexOf('*');
        string newExpression = string.Empty;
        newExpression = expression.Substring(0, index) + replacer + expression.Substring(index + 1);
        return newExpression.ToString();
    }

    static bool isVadidExpression(string expression)
    {        
        Stack<bool> validations = new Stack<bool>();
        for (int i = 0; i < expression.Length; i++)
        {
            if (expression[i] == '(')
            {
                validations.Push(true);
            }
            if (expression[i] == ')' && validations.Count > 0)
            {
                validations.Pop();
            }
            else if (expression[i] == ')' && validations.Count == 0)
            {
                return false;
            }
        }
        if (validations.Count == 0)
        {
            return true;
        }
        else
        {
            return false;
        }
    }
}
